File: framework/selective-connection/linux/core/src/ProtocolManager.cpp
/* Copyright 2017-2018 All Rights Reserved.
 *  Gyeonghwan Hong (redcarrottt@gmail.com)
 *  Eunsoo Park (esevan.park@gmail.com)
 *  Injung Hwang (sinban04@gmail.com)
 *
 * [Contact]
 *  Gyeonghwan Hong (redcarrottt@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0(the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../inc/ProtocolManager.h"

#include "../inc/SegmentManager.h"

#include "../../common/inc/DebugLog.h"

#include <netinet/in.h>
#include <stdlib.h>
#include <string.h>

using namespace sc;

uint16_t ProtocolManager::sPacketId = 1;

void ProtocolManager::data_to_protocol_data(const uint8_t *dat, uint32_t size,
                                            ProtocolData *ret_pd) {
  assert(ret_pd != NULL);
  assert(size <= UINT32_MAX);
  ret_pd->id = sPacketId++;
  ret_pd->len = size;
  ret_pd->data = dat;

#ifdef COMMUNICATOR_UNIT_TEST
  LOG_VERB("%d\t%u\t[%s]", ret_pd->id, ret_pd->len, ret_pd->data);
#endif /* COMMUNICATOR_UNIT_TEST */
}

inline void ProtocolManager::serialize_header(ProtocolData *pd,
                                              uint8_t *vec_ptr) {
  uint32_t vec_offset = 0;
  uint16_t net_id = htons(pd->id);
  uint32_t net_len = htonl(pd->len);

  memcpy(vec_ptr + vec_offset, &net_id, sizeof(uint16_t));
  vec_offset += sizeof(uint16_t);
  memcpy(vec_ptr + vec_offset, &net_len, sizeof(uint32_t));
  vec_offset += sizeof(uint32_t);
}

inline void ProtocolManager::serialize_data(const uint8_t *dat_buf,
                                            uint32_t len, uint8_t *vec_ptr) {
  memcpy(vec_ptr, dat_buf, (size_t)len);
}

uint32_t ProtocolManager::serialize(ProtocolData *pd, const uint8_t *buf,
                                    uint32_t offset, uint32_t payload_size,
                                    uint8_t **ret_vector) {
  uint8_t *serialized_vector;
  uint32_t vector_size;
  uint32_t data_offset;

  data_offset = kProtHeaderSize;
  vector_size = data_offset + payload_size;

  serialized_vector =
      reinterpret_cast<uint8_t *>(calloc((size_t)vector_size, sizeof(uint8_t)));

  serialize_header(pd, serialized_vector);
  serialize_data(buf + offset, payload_size, serialized_vector + data_offset);

#ifdef COMMUNICATOR_UNIT_TEST
  LOG_VERB("%p\t(%u)\tFrom %p~%p(%u)", serialized_vector, vector_size,
           buf + offset, buf + offset + payload_size - 1, payload_size);

  uint8_t buf_test = *(serialized_vector);
  buf_test = *(serialized_vector + vector_size - 1);
  if (buf != NULL) {
    buf_test = *(buf);
    buf_test = *(buf + offset + payload_size - 1);
  }
  LOG_VERB("Done");
#endif /* COMMUNICATOR_UNIT_TEST */

  *ret_vector = serialized_vector;

  return vector_size;
}

uint32_t ProtocolManager::parse_header(uint8_t *serialized,
                                       ProtocolData *ret_pd) {
  uint32_t vec_offset = 0;
  uint16_t net_id;
  uint32_t net_len;

  memcpy(&net_id, serialized + vec_offset, sizeof(uint16_t));
  vec_offset += sizeof(uint16_t);
  ret_pd->id = ntohs(net_id);

  memcpy(&net_len, serialized + vec_offset, sizeof(uint32_t));
  vec_offset += sizeof(uint32_t);
  ret_pd->len = ntohl(net_len);

  return vec_offset;
}

int ProtocolManager::send_packet(uint8_t *serialized, uint32_t packet_size, bool is_control) {
  SegmentManager *sm = SegmentManager::singleton();
#ifdef COMMUNICATOR_UNIT_TEST
  uint8_t buf_test = *serialized;
  buf_test = *(serialized + packet_size - 1);
  free(serialized);
#endif
  // Hand over the data to the segment manager
  return sm->send_to_segment_manager(serialized, packet_size, is_control);
}

uint32_t ProtocolManager::recv_packet(uint8_t **buf, bool is_control) {
  ProtocolData pd = {0, 0, NULL};
  SegmentManager *sm = SegmentManager::singleton();
  uint8_t *p_data_buf =
      sm->recv_from_segment_manager(reinterpret_cast<void *>(&pd), is_control);
  assert(pd.len == 0 || p_data_buf != NULL);

  *buf = p_data_buf;

  return pd.len;
}

File: subprojects/sc/android/app/src/main/java/com/redcarrottt/sc/internal/ProtocolManager.java
package com.redcarrottt.sc.internal;

/* Copyright (c) 2017-2018. All rights reserved.
 *  Gyeonghwan Hong (redcarrottt@gmail.com)
 *  Eunsoo Park (esevan.park@gmail.com)
 *  Injung Hwang (sinban04@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import java.nio.ByteBuffer;

class ProtocolData {
    short id;
    int len;
    byte[] data;

    ProtocolData() {
        id = 0;
        len = 0;
        data = null;
    }
}

class ProtocolManager {
    static public final int kProtocolHeaderSize = 6;

    static private short sPacketId;
    static private byte[] sSerializedVector = null;

    static private void serialize_header(ProtocolData pd, byte[] vec_ptr) {
        if (vec_ptr == null) throw new AssertionError();

        int vec_offset = 0;

        ByteBuffer buffer = ByteBuffer.allocate(2);

        buffer.putShort(pd.id);
        byte[] net_id = buffer.array();

        buffer = ByteBuffer.allocate(4);
        buffer.putInt(pd.len);
        byte[] net_len = buffer.array();

        System.arraycopy(net_id, 0, vec_ptr, vec_offset, 2);
        vec_offset += 2;

        System.arraycopy(net_len, 0, vec_ptr, vec_offset, 4);
    }

    static public ProtocolData data_to_protocol_data(byte[] buf, int len) {
        ProtocolData pd = new ProtocolData();

        pd.id = sPacketId++;
        pd.len = len;
        pd.data = buf;

        return pd;
    }

    static public int serialize(ProtocolData protocolData, byte[] buf, int offset, int
            payload_size) {
        int vec_size;
        int data_offset;

        data_offset = kProtocolHeaderSize;
        vec_size = data_offset + payload_size;

        sSerializedVector = new byte[vec_size];

        serialize_header(protocolData, sSerializedVector);
        System.arraycopy(buf, offset, sSerializedVector, data_offset, payload_size);

        if (sSerializedVector == null) throw new AssertionError();

        return vec_size;
    }

    static void parse_header(byte[] serialized, ProtocolData protocolData) {
        if (serialized == null || protocolData == null) throw new AssertionError();

        int vec_offset = 0;
        ByteBuffer buffer = ByteBuffer.allocate(2);

        buffer.put(serialized, vec_offset, 2);
        protocolData.id = buffer.getShort(0);
        vec_offset += 2;

        buffer = ByteBuffer.allocate(4);
        buffer.put(serialized, vec_offset, 4);
        protocolData.len = buffer.getInt(0);
    }

    static public int send_packet(int packetSize, boolean isControl) {
        SegmentManager sm = SegmentManager.singleton();

        return sm.send_to_segment_manager(sSerializedVector, packetSize, isControl);
    }

    static public int recv_packet(byte[] buf, boolean isControl) {
        if (buf == null) throw new AssertionError();

        ProtocolData protocolData = new ProtocolData();
        SegmentManager sm = SegmentManager.singleton();

        byte[] data = sm.recv_from_segment_manager(protocolData, isControl);

        System.arraycopy(data, 0, buf, 0, (protocolData.len < buf.length) ? protocolData.len :
                buf.length);


        return protocolData.len;
    }

//    private static String kTag = "protocol manager";
}